### ISSUE TITLE
Query builder WHERE clause wrongfully assumes IS NULL case

### ISSUE THREAD

**Issue by jorqensen** [laravel/framework#42743]
- Laravel Version: 9.17.0
- PHP Version: 8.1.6
- Database Driver & Version: MySQL 8.0.29 / SQLite 3.38.5 (Relateable to Postgres and other drivers as well.)

### Description:

Currently the query builder `WHERE` clause assumes if the `$value` is null, that the developer wants to do a `IS NULL` check. However, this can sometimes yield unexpected results.

In supported drivers, the spaceship operator (`<=>`) exists and is [NULL-safe equal to](https://dev.mysql.com/doc/refman/8.0/en/comparison-operators.html#operator_equal-to), while the `=` operator is not null safe.

Assume you have to do a query for a database lookup to find where `name is equal to but not null`, the query builder is intuitive enough to chain a `whereNotNull` and `where`. However, if you have to apply this logic for multiple fields, it can quickly become a mess.

With **valid** SQL we can just make use of the `=` operator and yield the same result without the multiple `NOT NULL` cases.

### Steps To Reproduce:

[I've created a demo repository with data that can be seeded and explanation + output example in routes.](https://github.com/jorqensen/laravel-where-clause-bug)

Otherwise:

1. Create fresh Laravel project
2. Make a nullable field, e.g. the `name` column in the users table.
3. Seed some data where name is `null` sometimes.
4. Run the below query:

```php
$name = "John Doe"; // Swap this with NULL

User::query()->where('name', '=', $name)->get();
```

### Proposed/expected behavior

In case the developer does not explicitly set the `WHERE` clause operator, I believe it's fine to assume `IS NULL`. But in case it's explicitly set, assumption should not happen.

Example:

```php
User::query()->where('name', null)->toSql();
// select * from `users` where `name` is null

User::query()->where('name', '=', null)->toSql();
// select * from `users` where `name` = null
```

[Here's the code that makes the assumption under the hood](https://github.com/laravel/framework/blob/9.x/src/Illuminate/Database/Query/Builder.php#L753). Below addition would "fix" it at a glance, I am unsure if this introduces some other breaking changes however.

```php
if (is_null($value) && is_null($operator)) {
    return $this->whereNull($column, $boolean, $operator !== '=');
}
```

**Comment by driesvints** [comment]
The current behavior is the correct expected behavior. You need to check the value up front like you do in your example.

**Comment by matiaslauriti** [comment]
> Out of curiosity, could you elaborate on why this is the expected/desired behavior of the query builder @driesvints ? 🙂
> 
> The point is just that `WHERE col IS NULL` is not the same as `WHERE col = NULL`. I find it odd that Laravel will by assumption throw my explicitly defined operator out the window when the value is `NULL`.

I am lost as what the issue is. When you run a query (at least for MySQL driver) you must do `WHERE column_name IS NULL` or `WHERE column_name IS NOT NULL`, doing `WHERE column_name = NULL` is not valid, so what is the issue? If Laravel receives `->where('column', null)` the only way to do that is using `IS NULL`

**Comment by jorqensen** [comment]
@matiaslauriti As explained and demonstrated in the example, `=` is not null safe while `<=>` is, they are both equality operators and perfectly valid to assert against `NULL` - they just yield opposite results.

If you have nullable field that you want to query against (or multiple) from something that isn't dictated beforehand, could be looping over a CSV, user input, etc. And you want to grab records `WHERE x = ?` but not grab records where the value is `NULL`, you can do the following:

```sql
select * from `users` where `name` = ?
```

If the value is not null, it will try and find records equal to or omit, as expected. In case of null, nothing is returned, as it's not null safe. That is also expected behavior.

The query builder will *always assume* that I want to assert `IS NULL` if my `$value` is `NULL`. Here:

```php
User::query()->where('name', '=', null)->get();
// select * from `users` where `name` is null
```

But this assumption is doing the exact *opposite* of the result you're trying to achieve.

My thought is more that, if I am explicit on my operator, Laravel should not be assuming what I am *actually* trying to do, as I clearly stated it. If it's a design choice that won't be changed, that's fine and I'll do raw queries, but one can be curious to why it's as such when it can lead to completely flipping your query on it's head.

**Comment by matiaslauriti** [comment]
> @matiaslauriti As explained and demonstrated in the example, `=` is not null safe while `<=>` is, they are both equality operators and perfectly valid to assert against `NULL` - they just yield opposite results.
> 
> If you have nullable field that you want to query against (or multiple) from something that isn't dictated beforehand, could be looping over a CSV, user input, etc. And you want to grab records `WHERE x = ?` but not grab records where the value is `NULL`, you can do the following:
> 
> ```sql
> select * from `users` where `name` = ?
> ```
> 
> If the value is not null, it will try and find records equal to or omit, as expected. In case of null, nothing is returned, as it's not null safe. That is also expected behavior.
> 
> The query builder will _always assume_ that I want to assert `IS NULL` if my `$value` is `NULL`. Here:
> 
> ```
> User::query()->where('name', '=', null)->get();
> // select * from `users` where `name` is null
> ```
> 
> But this assumption is doing the exact _opposite_ of the result you're trying to achieve.
> 
> My thought is more that, if I am explicit on my operator, Laravel should not be assuming what I am _actually_ trying to do, as I clearly stated it. If it's a design choice that won't be changed, that's fine and I'll do raw queries, but one can be curious to why it's as such when it can lead to completely flipping your query on it's head.

What I do not understand is why would I use `where name <=> null` instead of `is null`? I have never saw someone use that (I am not DBA), but just curious, why would Laravel need to change that behaviour? What value would it add?

**Comment by jorqensen** [comment]
@matiaslauriti I didn't say that, I just threw it there as an example that they're opposite, and they're opposite for a reason.

The value it would add IMO is that you would actually get back what you're writing out, instead of Laravel making an assumption flipping your query on the head.

To be clear, I do _**absolutely in no way**_ want to prevent Laravel from assuming `is null` when doing `->where('column', null)`. However, I do believe it would be more expected behavior that it _**did not**_ assume it when I _**explicitly**_ define my operator as: `->where('column', '=', null)`.

Because, again, `IS NULL` and `= NULL` will yield completely opposite results.

The goal of pointing this out isn't really debating wether or not if one should do X, Y or Z. It's to point out that it's a completely valid query that asserts a given case and yields expected results. But the default query builder where clause logic disallows one from doing as such without writing a raw query.

**Comment by matiaslauriti** [comment]
> @matiaslauriti I didn't say that, I just threw it there as an example that they're opposite, and they're opposite for a reason.
> 
> The value it would add IMO is that you would actually get back what you're writing out, instead of Laravel making an assumption flipping your query on the head.
> 
> To be clear, I do _**absolutely in no way**_ want to prevent Laravel from assuming `is null` when doing `->where('column', null)`. However, I do believe it would be more expected behavior that it _**did not**_ assume it when I _**explicitly**_ define my operator as: `->where('column', '=', null)`.
> 
> Because, again, `IS NULL` and `= NULL` will yield completely opposite results.
> 
> The goal of pointing this out isn't really debating whether or not if one should do X, Y or Z. It's to point out that it's a completely valid query that asserts a given case and yields expected results. But the default query builder where clause logic disallows one from doing as such without writing a raw query.

I still do not see the point, why would someone do `= null`? I have never saw that on SQL (MySQL, MariaDB). Also, remember that not passing `=` (`where($column, $value)`) will, by default, mean `=`. If `where($column, '=', null)` would be the only one having a different behavior, that would be break a lot of current running Laravels (let's say this is a new behaviour, it would be released on Laravel 10.x because it would break 9.x).

I am also asking out of curiosity, not debating if it is better or not, I just never saw that query and I am very intrigued and I can't find a lot of topics about it (with examples). Also, remember, that this would be a "MySQL" behaviour, as another SQL or NoSQL driver could have no `IS NULL` vs `<=> NULL` vs `= NULL`, so it would be very specific to MySQL and any similar engine

**Comment by jorqensen** [comment]
@matiaslauriti The primary reason this came to me was that I had to perform such a query at work the other day and was surprised how Laravel would *always* force the `IS NULL` case granted the value is `NULL`. As to why, that would take to long to explain and probably hit some NDA issues.

In short, the query needed to assert some values were equal to a string, but in some cases it would be null where we did not want a match. And we wanted to throw that "load" on the DB side.

Solved with a raw query, but I still believe the hard assumption of `IS NULL` in all cases isn't ideal.

I am aware it would most likely be a breaking change for some people, but one could also argue it would actually do what the developer *asked for* and not what Laravel *thinks* you want to do. I am on team Laravel with a lot of opinionated things, but seeing the query builder assuming as such with no way to override it caught me by surprise. It *feels* correct that if you are explicit with defining the operator, *that's* the operator in use.

With that being said, I'll stick to raw queries and accept this is "the correct expected behavior" - agree to disagree 🤷‍♂️
